README.md
# rework-calc [![Build Status](https://travis-ci.org/reworkcss/rework-calc.png)](https://travis-ci.org/reworkcss/rework-calc)

A [Rework](https://github.com/reworkcss/rework) plugin to support `calc()`.  
Particularly useful with the [rework-vars](https://github.com/reworkcss/rework-vars)

## Installation

```bash
npm install rework-calc
```

## Use

As a Rework plugin:

```javascript
var rework = require('rework'),
    calc = require('rework-calc');

var css = rework(cssString).use(calc).toString();
```

## Supported feature

This simply add `calc()` support, a feature to do simple calculations.
This can be particularly useful with the [rework-vars](https://github.com/reworkcss/rework-vars) plugin.

**Note:** When multiple units are mixed together in the same expression, the `calc()` statement
is left as is, to fallback to the CSS3 calc feature.

**Example** (with [rework-vars](https://github.com/reworkcss/rework-vars) enabled as well):

```css
:root {
  --main-font-size: 16px;
}

body {
  font-size: var(--main-font-size);
}

h1 {
  font-size: calc(var(--main-font-size) * 2);
  height: calc(100px - 2em);
}
```

**yields**:

```css
body {
  font-size: 16px
}

h1 {
  font-size: 32px;
  height: calc(100px - 2em)
}
```

See unit tests for another example.

## Unit tests

Make sure the dev-dependencies are installed, and then run:

```bash
npm test
```

## License

MIT


History.md
1.1.0 / 2014-07-05
==================

* resolution of nested calc()

1.0.0 / 2014-06-25
==================

* error reporting

0.2.2 / 2014-05-03
==================

* take care of vendor prefixed expressions

0.2.1 / 2013-12-13
==================

 * first standalone release


test/test.js
var assert = require('assert');
var calc = require('..');
var read = require('fs').readFileSync;
var rework = require('rework');
var vars = require('rework-vars');

function fixture(name) {
  return read('test/fixtures/' + name + '.css', 'utf8').trim();
}

function compareFixtures(name) {
  var actual = rework(fixture(name + '.in')).use(calc).toString().trim();
  var expected = fixture(name + '.out');
  return assert.equal(actual, expected);
}

describe('rework-calc', function () {
  it('throws an error when a calc function is empty', function () {
    var output = function () {
      return rework(fixture('substitution-empty')).use(calc).toString();
    };
    assert.throws(output, Error, 'rework-calc: calc() must contain a non-whitespace string');
  });

  it('throws an error when a calc function is malformed', function () {
    var output = function () {
      return rework(fixture('substitution-malformed')).use(calc).toString();
    };
    assert.throws(output, Error, 'rework-calc: missing closing ")" in the value "calc(10px - 5px"');
  });

  it('should calculate expressions with only one unit involved', function () {
    compareFixtures('calc');
  });

  it('should calculate expressions with percents correctly', function () {
    compareFixtures('calc-percent');
  });

  it('should use CSS3 Calc function as fallback for expressions with multiple units', function () {
    compareFixtures('calc-fallback');
  });

  it('should handle vendor prefixed expressions', function () {
    compareFixtures('calc-prefix');
  });

  it('should resolve what is possible in complex calc whenever it\'s possible', function () {
    var name = 'calc-complex'
    var actual = rework(fixture(name + '.in'))
      .use(vars())
      .use(calc)
      .toString()
      .trim();
    var expected = fixture(name + '.out');
    return assert.equal(actual, expected);
  });
});


test/plugins.js
var calc = require('../index'),
    rework = require('rework'),
    should = require('chai').Should(),
    read = require('fs').readFileSync;

var css = {
  in: function (name) {
    return this._read(name, 'in');
  },
  out: function (name) {
    return this._read(name, 'out');
  },
  _read: function (name, type) {
    return read(__dirname + '/' + name + '.' + type + '.css', 'utf8');
  }
};

describe('rework-calc', function() {
  it('should calculate expressions with only one unit involved', function() {
    rework(css.in('calc'))
      .use(calc)
      .toString()
      .should.equal(css.out('calc'));
  });

  it('should calculate expressions with percents correctly', function () {
    rework(css.in('calc-percent'))
      .use(calc)
      .toString()
      .should.equal(css.out('calc-percent'));
  });

  it('should use CSS3 Calc function as fallback for expressions with multiple units', function () {
    rework(css.in('calc-complex'))
      .use(calc)
      .toString()
      .should.equal(css.out('calc-complex'));
  });

  it('should handle vendor prefixed expressions', function () {
    rework(css.in('calc-prefix'))
      .use(calc)
      .toString()
      .should.equal(css.out('calc-prefix'));
  });

});


index.js
/**
 * Module dependencies.
 */

var balanced = require('balanced-match');
var visit = require('rework-visit');

/**
 * Constants.
 */

var DEFAULT_UNIT = 'px';
var CALC_FUNC_IDENTIFIER =  'calc';
var EXPRESSION_OPT_VENDOR_PREFIX = '(\\-[a-z]+\\-)?';
var EXPRESSION_METHOD_REGEXP = EXPRESSION_OPT_VENDOR_PREFIX + CALC_FUNC_IDENTIFIER;
var EXPRESSION_REGEXP = '\\b' + EXPRESSION_METHOD_REGEXP + '\\(';

/**
 * Module export.
 */

module.exports = function calc(style) {
  // resolve calculations
  visit(style, function (declarations, node) {
    var decl;
    var resolvedValue;
    var value;

    for (var i = 0; i < declarations.length; i++) {
      decl = declarations[i];
      value = decl.value;

      // skip comments
      if (decl.type !== 'declaration') continue;
      // skip values that don't contain calc() functions
      if (!value || value.indexOf(CALC_FUNC_IDENTIFIER + '(') === -1) continue;

      decl.value = resolveValue(value);
    }
  });
};

/**
 * Parses expressions in a value
 *
 * @param {String} value
 * @returns {Array}
 * @api private
 */

function getExpressionsFromValue(value) {
  var parentheses = 0;
  var expressions = [];
  var start = null;

  // Parse value and extract expressions:
  for (var i = 0; i < value.length; i++) {
    if (value[i] == '(' && value.slice(i - 4, i) == CALC_FUNC_IDENTIFIER && !start) {
      start = i;
      parentheses++;
    } else if (value[i] == '(' && start !== null) {
      parentheses++;
    } else if (value[i] == ')' && start !== null) {
      if (!--parentheses) {
        expressions.push(value.slice(start + 1, i));
        start = null;
      }
    }
  }

  return expressions;
}

/**
 * Walkthrough all expressions, evaluate them and insert them into the declaration
 *
 * @param {Array} expressions
 * @param {Object} declaration
 * @api private
 */

function resolveValue(value) {
  var balancedParens = balanced('(', ')', value);
  var calcStartIndex = value.indexOf(CALC_FUNC_IDENTIFIER + '(');
  var calcRef = balanced('(', ')', value.substring(calcStartIndex));

  if (!balancedParens) throw new Error('rework-calc: missing closing ")" in the value "' + value + '"');
  if (!calcRef || calcRef.body === '') throw new Error('rework-calc: calc() must contain a non-whitespace string');

  getExpressionsFromValue(value).forEach(function (expression) {
    var result = evaluateExpression(expression);

    if (!result) return;

    // Insert the evaluated value:
    var expRegexp = new RegExp(EXPRESSION_METHOD_REGEXP + '\\(' + escapeExp(expression) + '\\)');
    value = value.replace(expRegexp, result);
  });

  return value;
}

/**
 * Evaluates an expression
 *
 * @param {String} expression
 * @returns {String}
 * @api private
 */

function evaluateExpression (expression) {
  var originalExpression = CALC_FUNC_IDENTIFIER + '(' + expression + ')';

  // Remove method names for possible nested expressions:
  expression = expression.replace(new RegExp(EXPRESSION_REGEXP, 'g'), '(');

  var uniqueUnits = getUnitsInExpression(expression);

  // If multiple units let the expression be (i.e. fallback to CSS3 calc())
  if (uniqueUnits.length > 1) return false;

  if (!uniqueUnits.length) {
    console.warn('No unit found in expression: "' + originalExpression + '", defaults to: "' + DEFAULT_UNIT + '"');
  }

  // Use default unit if needed:
  var unit = uniqueUnits[0] || DEFAULT_UNIT;

  if (unit === '%') {
    // Convert percentages to numbers, to handle expressions like: 50% * 50% (will become: 25%):
    expression = expression.replace(/\b[0-9\.]+%/g, function (percent) {
      return parseFloat(percent.slice(0, -1)) * 0.01;
    });
  }

  // Remove units in expression:
  var toEvaluate = expression.replace(new RegExp(unit, 'g'), '');

  var result;

  try {
    result = eval(toEvaluate);
  } catch (e) {
    return false;
  }

  // Transform back to a percentage result:
  if (unit === '%') result *= 100;

  // We don't need units for zero values...
  if (result !== 0) result += unit;

  return result;
}

/**
 * Checks what units are used in an expression
 *
 * @param {String} expression
 * @returns {Array}
 * @api private
 */

function getUnitsInExpression(expression) {
  var uniqueUnits = [];
  var unitRegEx = /[\.0-9]([%a-z]+)/g;
  var matches;

  while (matches = unitRegEx.exec(expression)) {
    if (!matches || !matches[1]) continue;
    if (!~uniqueUnits.indexOf(matches[1])) uniqueUnits.push(matches[1]);
  }

  return uniqueUnits;
}

/**
 * Escape string for inclusion in a regex pattern without conflicts
 *
 * @param  {String} str String to escape
 * @return {String} Escaped string
 * @api private
 */

function escapeExp(str) {
  return String(str).replace(/([.*+?^=!:${}()|\[\]\/\\])/g, '\\$1');
}


lib/calc.js
/**
 * Calculation Plugin
 *
 * Useful in combination with the [rework-vars](https://npmjs.org/package/rework-vars) plugin, e.g:
 *
 *   :root {
 *     --base-font-size: 16px;
 *   }
 *   body {
 *     font-size: var(--base-font-size);
 *   }
 *   h1 {
 *     font-size: calc(var(--base-font-size) * 2);
 *   }
 *
 * Yields:
 *
 *   body {
 *     font-size: 16px;
 *   }
 *   h1 {
 *     font-size: 32px;
 *   }
 *
 */

/**
 * Module dependencies.
 */

var balanced = require('balanced-match');
var visit = require('rework-visit');

/**
 * Constants.
 */
var DEFAULT_UNIT = 'px',
    CALC_FUNC_IDENTIFIER =  'calc',

    EXPRESSION_OPT_VENDOR_PREFIX = '(\\-[a-z]+\\-)?',
    EXPRESSION_METHOD_REGEXP = EXPRESSION_OPT_VENDOR_PREFIX + CALC_FUNC_IDENTIFIER,
    EXPRESSION_REGEXP = '\\b' + EXPRESSION_METHOD_REGEXP + '\\(';

/**
 * Module export.
 */

module.exports = function (options) {

  return function vars(style) {
    options = options || {};
    var preserve = (options.preserve === true ? true : false);

    // resolve variables
    visit(style, function (declarations, node) {
      var decl;
      var resolvedValue;
      var value;

      for (var i = 0; i < declarations.length; i++) {
        decl = declarations[i];
        value = decl.value;

        // skip comments
        if (decl.type !== 'declaration') continue;
        // skip values that don't contain variable functions
        if (!value || value.indexOf(CALC_FUNC_IDENTIFIER + '(') === -1) continue;

        resolvedValue = resolveValue(value);

        if (!preserve) {
          decl.value = resolvedValue;
        }
        else {
          declarations.splice(i, 0, {
            type: decl.type,
            property: decl.property,
            value: resolvedValue
          });
          // skip ahead of preserved declaration
          i++;
        }
      }
    });
  };
};

/**
 * Parses expressions in a value
 *
 * @param {String} value
 * @returns {Array}
 * @api private
 */
function getExpressionsFromValue(value) {
    var parentheses = 0,
        expressions = [],
        start = null;

    // Parse value and extract expressions:
    for (var i = 0; i < value.length; i++) {
        if (value[i] == '(' && value.slice(i - 4, i) == CALC_FUNC_IDENTIFIER && !start) {
            start = i;
            parentheses++;
        } else if (value[i] == '(' && start !== null) {
            parentheses++;
        } else if (value[i] == ')' && start !== null) {
            if (!--parentheses) {
                expressions.push(value.slice(start + 1, i));
                start = null;
            }
        }
    }

    return expressions;
}

/**
 * Walkthrough all expressions, evaluate them and insert them into the declaration
 *
 * @param {Array} expressions
 * @param {Object} declaration
 * @api private
 */
function resolveValue(value) {
  var balancedParens = balanced('(', ')', value);
  var calcRef = balanced(CALC_FUNC_IDENTIFIER + '(', ')', value);

  if (!balancedParens) throw new Error('rework-calc: missing closing ")" in the value "' + value + '"');
  if (!calcRef || calcRef.body === '') throw new Error('rework-calc: calc() must contain a non-whitespace string');

  getExpressionsFromValue(value).forEach(function (expression) {
    var result = evaluateExpression(expression);

    if (!result) return;

    // Insert the evaluated value:
    var expRegexp = new RegExp(EXPRESSION_METHOD_REGEXP + '\\(' + escapeExp(expression) + '\\)');
    value = value.replace(expRegexp, result);
  });

  return value
}

/**
 * Evaluates an expression
 *
 * @param {String} expression
 * @returns {String}
 * @api private
 */
function evaluateExpression (expression) {
    var originalExpression = CALC_FUNC_IDENTIFIER + '(' + expression + ')';

    // Remove method names for possible nested expressions:
    expression = expression.replace(new RegExp(EXPRESSION_REGEXP, 'g'), '(');

    var uniqueUnits = getUnitsInExpression(expression);

    // If multiple units let the expression be (i.e. fallback to CSS3 calc())
    if (uniqueUnits.length > 1)
        return false;

    if (!uniqueUnits.length)
        console.warn('No unit found in expression: "' + originalExpression + '", defaults to: "' + DEFAULT_UNIT + '"');

    // Use default unit if needed:
    var unit = uniqueUnits[0] || DEFAULT_UNIT;

    if (unit === '%') {
        // Convert percentages to numbers, to handle expressions like: 50% * 50% (will become: 25%):
        expression = expression.replace(/\b[0-9\.]+%/g, function (percent) {
            return parseFloat(percent.slice(0, -1)) * 0.01;
        });
    }

    // Remove units in expression:
    var toEvaluate = expression.replace(new RegExp(unit, 'g'), '');

    var result;

    try {
        result = eval(toEvaluate);
    } catch (e) {
        return false;
    }

    // Transform back to a percentage result:
    if (unit === '%')
        result *= 100;

    // We don't need units for zero values...
    if (result !== 0)
        result += unit;

    return result;
}

/**
 * Checks what units are used in an expression
 *
 * @param {String} expression
 * @returns {Array}
 * @api private
 */
function getUnitsInExpression(expression) {
    var uniqueUnits = [],
        unitRegEx = /[\.0-9]([%a-z]+)/g,
        matches;

    while (matches = unitRegEx.exec(expression)) {
        if (!matches || !matches[1]) continue;
        if (!~uniqueUnits.indexOf(matches[1]))
            uniqueUnits.push(matches[1]);
    }

    return uniqueUnits;
}

/**
 * Escape string for inclusion in a regex pattern without conflicts
 *
 * @param  {String} str String to escape
 * @return {String} Escaped string
 * @api private
 */
function escapeExp(str) {
    return String(str).replace(/([.*+?^=!:${}()|\[\]\/\\])/g, '\\$1');
}
